GPU/TPCFastTransformation/Spline1D.cxx

```diff
@@ -169,48 +205,70 @@ void Spline1D::constructKnotsRegular(int numberOfKnots)
   for (int i = 0; i < numberOfKnots; i++) {
     knots[i] = i;
   }
-  constructKnots(numberOfKnots, knots.data());
+  recreate(numberOfKnots, knots.data(), nFdimensions);
 }
+
 #endif
 
-void Spline1D::print() const
+template <typename DataT>
+void Spline1D<DataT>::print() const
 {
-#if !defined(GPUCA_GPUCODE)
-  std::cout << " Compact Spline 1D: " << std::endl;
-  std::cout << "  mNumberOfKnots = " << mNumberOfKnots << std::endl;
-  std::cout << "  mUmax = " << mUmax << std::endl;
-  std::cout << "  mUtoKnotMap = " << (void*)mUtoKnotMap << std::endl;
-  std::cout << "  knots: ";
+  printf(" Compact Spline 1D: \n");
+  printf("  mNumberOfKnots = %d \n", mNumberOfKnots);
+  printf("  mUmax = %d\n", mUmax);
+  printf("  mUtoKnotMap = %p \n", (void*)mUtoKnotMap);
+  printf("  knots: ");
   for (int i = 0; i < mNumberOfKnots; i++) {
-    std::cout << getKnot(i).u << " ";
+    printf("%d ", (int)getKnot(i).u);
   }
-  std::cout << std::endl;
-#endif
+  printf("\n");
 }
 
-#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE) // code invisible on GPU and in the standalone compilation
+#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE)
+
+template <typename DataT>
+void Spline1D<DataT>::approximateFunction(DataT xMin, DataT xMax,
+                                          std::function<void(DataT x, DataT f[/*mFdimensions*/])> F,
+                                          int nAxiliaryDataPoints)
+{
+  /// Approximate F with this spline
+  setXrange(xMin, xMax);
+  SplineHelper1D<DataT> helper;
+  helper.approximateFunction(*this, xMin, xMax, F, nAxiliaryDataPoints);
+}
+
+template <typename DataT>
+int Spline1D<DataT>::writeToFile(TFile& outf, const char* name)
+{
+  /// write a class object to the file
+  return FlatObject::writeToFile(*this, outf, name);
+}
 
-int Spline1D::test(const bool draw, const bool drawDataPoints)
+template <typename DataT>
+Spline1D<DataT>* Spline1D<DataT>::readFromFile(TFile& inpf, const char* name)
+{
+  /// read a class object from the file
+  return FlatObject::readFromFile<Spline1D<DataT>>(inpf, name);
+}
+
+template <typename DataT>
+int Spline1D<DataT>::test(const bool draw, const bool drawDataPoints)
 {
   using namespace std;
 
-  const int Ndim = 5;
+  // input function F
 
+  const int Ndim = 5;
   const int Fdegree = 4;
-
   double Fcoeff[Ndim][2 * (Fdegree + 1)];
 
-  int nKnots = 4;
-  const int nAxiliaryPoints = 1;
-  int uMax = nKnots * 3;
-
-  auto F = [&](float u, float f[]) -> void {
-    double uu = u * TMath::Pi() / uMax;
+  auto F = [&](DataT x, DataT f[]) -> void {
+    double xx = x;
     for (int dim = 0; dim < Ndim; dim++) {
       f[dim] = 0; // Fcoeff[0]/2;
       for (int i = 1; i <= Fdegree; i++) {
-        f[dim] += Fcoeff[dim][2 * i] * TMath::Cos(i * uu) +
-                  Fcoeff[dim][2 * i + 1] * TMath::Sin(i * uu);
+        f[dim] += Fcoeff[dim][2 * i] * TMath::Cos(i * xx) +
```

Comment:
should not be more efficient
````
double xx = 0;
for (int dim = 0; dim < Ndim; dim++) {
  f[dim] = 0; // Fcoeff[0]/2;
    for (int i = 1; i <= Fdegree; i++) {
      xx += x;
      f[dim] += Fcoeff[dim][2 * i] * TMath::Cos(xx) + Fcoeff[dim][2 * i + 1] * TMath::Sin(xx); // eventually, use sincos ?
    }
}

Reply:
Sure, this way it is faster.  I didn't care about performance, because it is just a test method.
But, why not:)  I've changed the code, also in Spline2D::test().

Reply:
2 @shahor02:  By the way, sincos() doesn't exist on Mac OsX, the compilation crashes.

Reply:
Can you try __sincos on Mac?

Reply:
I don't have my own Mac installation, therefore I can't check it locally.  
And I can't add __sincos to the PR, since it doesn't compile on CentOS. 
Is there an #ifdef for macosx?

Reply:
You could try
```
#ifdef __APPLE__
#define sincos(x, s, c) __sincos(x, s, c)
#define sincosf(x, s, c) __sincosf(x, s, c)
#endif
```

Reply:
@davidrohr: it is anyhow just a test method, its speed is not important.

Reply:
@sgorbuno understood that this is just a test, just for the record: I am collecting some methods in a separate namespace, including also the non-portable sincos: https://github.com/AliceO2Group/AliceO2/blob/a0f76c0489b9f59b81b4814927b4367f5df3a5a3/Common/MathUtils/include/MathUtils/Utils.h#L70. At the moment with separate sin and cos calls, eventually, will add case by case implementations. You are welcome to complete these mathutils as needed.

Reply:
@shahor02 : David's patch for the macOsX works. I'll port it to your Utils.h, ok?

Reply:
sure, can also add at once the GNU extension?
````
#ifdef __GNU_SOURCE__
 sincosf(ang,&s,&c);
#endif
````

Reply:
I've put the sincos() calls to GPU/Common/GPUCommonMath.h:

GPUhdi() void GPUCommonMath::SinCos(float x, float& s, float& c)
{
#if defined(__APPLE__)
  __sincosf(x, &s, &c);
#elif defined(__GNU_SOURCE__)
  sincosf(x, &s, &c);
#else
  CHOICE({s = sin(x); c = cos(x); }, sincosf(x, &s, &c), s = sincos(x, &c));
#endif
}

Reply:
AliceO2/Common/MathUtils/include/MathUtils/Utils.h calls gpu::CommonMath::sincos()

Reply:
Thanks!